src/transform/src/predicate_pushdown.rs

```diff
@@ -607,6 +619,69 @@ impl PredicatePushdown {
         }
     }
 
+    /// If `s` is of the form
+    /// `(isnull(expr1) && isnull(expr2)) || (expr1 = expr2)`,
+    /// extract `expr1` and `expr2`.
+    fn extract_equal_or_both_null(s: &mut MirScalarExpr) -> Option<(MirScalarExpr, MirScalarExpr)> {
+        // Or, And, and Eq are all commutative functions. For each of these
+        // functions, order expr1 and expr2 so you only need to check
+        // `condition1(expr1) && condition2(expr2)`, and you do
+        // not need to also check for `condition2(expr1) && condition1(expr2)`.
```

Comment:
This branch has an extra notation on the ```Join``` line of the plan that is not present in the previous commit. Is that to be expected?

```
materialize=> EXPLAIN SELECT  * FROM t1 RIGHT JOIN t2 ON ( TRUE ) WHERE t2.f2 = t1.f2 AND t1.f2 = t2.f1 + t2.f2;
                Optimized Plan                
----------------------------------------------
 %0 =                                        +
 | Get materialize.public.t1 (u57)           +
 | ArrangeBy (#1)                            +
                                             +
 %1 =                                        +
 | Get materialize.public.t2 (u61)           +
 | Filter (#1 = (#0 + #1))                   +
                                             +
 %2 =                                        +
 | Join %0 %1 (= #1 #3 (#2 + #1))            + <- This part "(#2 + #1)"
 | | implementation = Differential %1 %0.(#1)+
 | | demand = (#0..#2)                       +
 | Project (#0..#2, #1)                      +
```

Reply:
This has been fixed. The problem was that I did not delete the predicates from the body of a let after pushing them down, so they would get re-pushed into a join condition.